File: packages/core/src/recurrence.ts
export type Recurrence = "daily" | "weekdays" | "weekly" | "monthly";

export const RECURRENCE_LABELS: Record<Recurrence, string> = {
  daily: "Daily",
  weekdays: "Weekdays",
  weekly: "Weekly",
  monthly: "Monthly",
};

function parseDay(iso: string): Date {
  return new Date(`${iso.slice(0, 10)}T00:00:00`);
}

function fmtDay(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

/** The next due date (YYYY-MM-DD) after `fromIso` for the given cadence. */
export function nextOccurrence(fromIso: string, recurrence: Recurrence): string {
  const d = parseDay(fromIso);
  switch (recurrence) {
    case "daily":
      d.setDate(d.getDate() + 1);
      break;
    case "weekdays":
      do {
        d.setDate(d.getDate() + 1);
      } while (d.getDay() === 0 || d.getDay() === 6);
      break;
    case "weekly":
      d.setDate(d.getDate() + 7);
      break;
    case "monthly": {
      // Clamp to the last day of the next month (Jan 31 -> Feb 28).
      const dayOfMonth = d.getDate();
      const lastOfNext = new Date(d.getFullYear(), d.getMonth() + 2, 0);
      d.setDate(1);
      d.setMonth(d.getMonth() + 1);
      d.setDate(Math.min(dayOfMonth, lastOfNext.getDate()));
      break;
    }
  }
  return fmtDay(d);
}

/**
 * Whether a task anchored at `anchorIso` (its due date) is scheduled on
 * `dayIso` under the given cadence.
 */
export function isScheduledOn(
  anchorIso: string,
  recurrence: Recurrence,
  dayIso: string,
): boolean {
  const anchor = parseDay(anchorIso);
  const day = parseDay(dayIso);
  switch (recurrence) {
    case "daily":
      return true;
    case "weekdays":
      return day.getDay() >= 1 && day.getDay() <= 5;
    case "weekly":
      return day.getDay() === anchor.getDay();
    case "monthly":
      return day.getDate() === anchor.getDate();
  }
}

File: apps/android/app/src/main/java/com/dahoko/android/domain/Recurrence.kt
package com.dahoko.android.domain

import java.time.DayOfWeek
import java.time.LocalDate
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class Recurrence(val label: String) {
    @SerialName("daily") DAILY("Daily"),
    @SerialName("weekdays") WEEKDAYS("Weekdays"),
    @SerialName("weekly") WEEKLY("Weekly"),
    @SerialName("monthly") MONTHLY("Monthly");

    val wireName: String get() = label.lowercase()

    companion object {
        fun fromWire(value: String?): Recurrence? =
            entries.firstOrNull { it.wireName == value }
    }
}

private fun parseDay(iso: String): LocalDate = LocalDate.parse(iso.take(10))

/** The next due date (YYYY-MM-DD) after [fromIso] for the given cadence. */
fun nextOccurrence(fromIso: String, recurrence: Recurrence): String {
    var d = parseDay(fromIso)
    when (recurrence) {
        Recurrence.DAILY -> d = d.plusDays(1)
        Recurrence.WEEKDAYS -> {
            do {
                d = d.plusDays(1)
            } while (d.dayOfWeek == DayOfWeek.SATURDAY || d.dayOfWeek == DayOfWeek.SUNDAY)
        }
        Recurrence.WEEKLY -> d = d.plusWeeks(1)
        Recurrence.MONTHLY -> {
            // Clamp to the last day of the next month (Jan 31 -> Feb 28).
            val dayOfMonth = d.dayOfMonth
            val nextMonth = d.withDayOfMonth(1).plusMonths(1)
            d = nextMonth.withDayOfMonth(minOf(dayOfMonth, nextMonth.lengthOfMonth()))
        }
    }
    return d.toString()
}

/**
 * Whether a task anchored at [anchorIso] (its due date) is scheduled on
 * [dayIso] under the given cadence.
 */
fun isScheduledOn(anchorIso: String, recurrence: Recurrence, dayIso: String): Boolean {
    val anchor = parseDay(anchorIso)
    val day = parseDay(dayIso)
    return when (recurrence) {
        Recurrence.DAILY -> true
        Recurrence.WEEKDAYS -> day.dayOfWeek != DayOfWeek.SATURDAY && day.dayOfWeek != DayOfWeek.SUNDAY
        Recurrence.WEEKLY -> day.dayOfWeek == anchor.dayOfWeek
        Recurrence.MONTHLY -> day.dayOfMonth == anchor.dayOfMonth
    }
}
